REGRESSION (UI-side compositing): Jumpy full screen transition animation

rr-codes opened the pull request:
#### 3d721c09585b37b63eb4172c53ab61da282d2c3e
<pre>
REGRESSION (UI-side compositing): Jumpy full screen transition animation
<a href="https://bugs.webkit.org/show_bug.cgi?id=259375">https://bugs.webkit.org/show_bug.cgi?id=259375</a>
rdar://111534888

Reviewed by Wenson Hsieh.

With UI-side compositing, when entering or exiting window fullscreen mode, the transition is
sometimes jumpy. This happens in cases where the UI process resizes before the web content has a
chance to resize. The web content and the UI Process should always resize in the same CA commit.

Fix by adopting an AppKit SPI to defer entering or exiting window fullscreen mode until the web
content has actually been resized to the new size. This is done by using a &quot;window snapshot readiness handler&quot;.

* Source/WTF/wtf/PlatformHave.h:
Add a new `HAVE` definition to only enable this on macOS, and only on versions where the SPI exists.

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView dealloc]):
(-[WKWebView _invalidateWindowSnapshotReadinessHandler]):
Invokes the readiness handler and then resets it, indicating to AppKit that the full screen should
be allowed to resume.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKView.mm:
(-[WKView _web_windowWillEnterFullScreen]):
(-[WKView _web_windowWillExitFullScreen]):
Stub methods required to compile legacy WebKit.

* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _doWindowSnapshotReadinessUpdate]):
(-[WKWebView setFrameSize:]):
When `setFrameSize` gets called with the new frame size, the UI process waits for the next presentation
update of the web prcoess. Once in the completion handler, we know that the web content has properly
resized, and so we should now invalidate the readiness handler.

(-[WKWebView _web_windowWillEnterFullScreen]):
(-[WKWebView _web_windowWillExitFullScreen]):
Upon receiving one of these notifications, we take out a window snapshot readiness handler and store
it. This tells AppKit to defer the final stage of the enter and exit full screen animations until
the handler is invoked.

* Source/WebKit/UIProcess/mac/WKViewLayoutStrategy.h:
* Source/WebKit/UIProcess/mac/WKViewLayoutStrategy.mm:
(-[WKViewLayoutStrategy disableFrameSizeUpdates]):
(-[WKViewLayoutStrategy enableFrameSizeUpdates]):
(-[WKViewLayoutStrategy frameSizeUpdatesDisabled]):
(-[WKViewLayoutStrategy didChangeFrameSize]):
In some cases, `setFrameSize` would early return inside of `didChangeFrameSize` due to these set of
methods. This caused the bug to still reproduce, since the new drawing area size was no longer being set.

These methods have long been not actually needed, so their implementations are now removed and there is
no longer a need for an early return.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKWindowVisibilityObserver startObserving:]):
(-[WKWindowVisibilityObserver stopObserving:]):
(-[WKWindowVisibilityObserver _windowWillEnterFullScreen:]):
(-[WKWindowVisibilityObserver _windowWillExitFullScreen:]):
(WebKit::WebViewImpl::windowWillEnterFullScreen):
(WebKit::WebViewImpl::windowWillExitFullScreen):
When entering or exiting full screen, we now listen to their respective notifications.

Canonical link: <a href="https://commits.webkit.org/266273@main">https://commits.webkit.org/266273@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73dda3a41afcb252a6a94e3885cf358e0ede1e8a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13385 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13699 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14031 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15122 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12741 "Failed to checkout and rebase branch from PR 15961") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16206 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13725 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15421 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13552 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14207 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11326 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15789 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12081 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19126 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11407 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12570 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12251 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15457 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12671 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12750 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10609 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13430 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12021 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3521 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3267 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16344 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13814 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12592 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3318 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/mac/WKViewLayoutStrategy.mm:
```diff
@@ -136,9 +136,6 @@ - (void)didEndLiveResize
 
 - (void)didChangeFrameSize
 {
-    if ([self frameSizeUpdatesDisabled])
```
You should remove all the code out from here, too.

rr-codes replied:
yup I know. This is just a draft PR for now until I fix everything up

shultzc commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1704,6 +1706,15 @@ - (void)_recalculateViewportSizesWithMinimumViewportInset:(CocoaEdgeInsets)minim
 
+- (void)_invalidateWindowSnapshotReadynessHandler
+{
+    if (!self->_windowSnapshotReadynessHandler)
```
Nit: Read**i**ness here and throughout

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1697,6 +1701,17 @@ - (void)_recalculateViewportSizesWithMinimumViewportInset:(CocoaEdgeInsets)minim
+        return;
+
+    self->_windowSnapshotReadinessHandler();
+    self->_windowSnapshotReadinessHandler = nil;
```
```suggestion
    if (auto handler = std::exchange(_windowSnapshotReadinessHandler, nil))
        handler();
```

rr-codes replied:
good idea!

pxlcoder commented on Source/WebKit/UIProcess/mac/WKViewLayoutStrategy.mm:
```diff
@@ -105,21 +105,15 @@ - (void)updateLayout
 - (BOOL)frameSizeUpdatesDisabled
 {
-    return _frameSizeUpdatesDisabledCount > 0;
+    return false;
```
```suggestion
    return NO;
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -139,6 +139,10 @@ struct TranslationContextMenuInfo;
 - (void)_web_didChangeContentSize:(NSSize)newSize;
 
+- (void)_web_windowWillEnterFullScreen;
+
```
Nit: Remove newline.

whsieh commented on Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
```diff
@@ -1177,6 +1195,26 @@ - (void)_web_didChangeContentSize:(NSSize)newSize
+        return;
+
+    _windowSnapshotReadinessHandler = makeBlockPtr([self.window _holdResizeSnapshotWithReason:@"web view entering full screen"]);
+#endif
```
Nit - you could probably factor this code out into a helper function like `-_holdWindowResizeSnapshotWithReasion:(NSString *)reason;`, to avoid a tiny bit of code duplication in `-_web_windowWillEnterFullScreen` and `-_web_windowWillExitFullScreen`

rr-codes committed:
REGRESSION (UI-side compositing): Jumpy full screen transition animation
https://bugs.webkit.org/show_bug.cgi?id=259375
rdar://111534888

Reviewed by Wenson Hsieh.

With UI-side compositing, when entering or exiting window fullscreen mode, the transition is
sometimes jumpy. This happens in cases where the UI process resizes before the web content has a
chance to resize. The web content and the UI Process should always resize in the same CA commit.

Fix by adopting an AppKit SPI to defer entering or exiting window fullscreen mode until the web
content has actually been resized to the new size. This is done by using a "window snapshot readiness handler".

* Source/WTF/wtf/PlatformHave.h:
Add a new `HAVE` definition to only enable this on macOS, and only on versions where the SPI exists.

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView dealloc]):
(-[WKWebView _invalidateWindowSnapshotReadinessHandler]):
Invokes the readiness handler and then resets it, indicating to AppKit that the full screen should
be allowed to resume.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKView.mm:
(-[WKView _web_windowWillEnterFullScreen]):
(-[WKView _web_windowWillExitFullScreen]):
Stub methods required to compile legacy WebKit.

* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _doWindowSnapshotReadinessUpdate]):
(-[WKWebView setFrameSize:]):
When `setFrameSize` gets called with the new frame size, the UI process waits for the next presentation
update of the web prcoess. Once in the completion handler, we know that the web content has properly
resized, and so we should now invalidate the readiness handler.

(-[WKWebView _web_windowWillEnterFullScreen]):
(-[WKWebView _web_windowWillExitFullScreen]):
Upon receiving one of these notifications, we take out a window snapshot readiness handler and store
it. This tells AppKit to defer the final stage of the enter and exit full screen animations until
the handler is invoked.

* Source/WebKit/UIProcess/mac/WKViewLayoutStrategy.h:
* Source/WebKit/UIProcess/mac/WKViewLayoutStrategy.mm:
(-[WKViewLayoutStrategy disableFrameSizeUpdates]):
(-[WKViewLayoutStrategy enableFrameSizeUpdates]):
(-[WKViewLayoutStrategy frameSizeUpdatesDisabled]):
(-[WKViewLayoutStrategy didChangeFrameSize]):
In some cases, `setFrameSize` would early return inside of `didChangeFrameSize` due to these set of
methods. This caused the bug to still reproduce, since the new drawing area size was no longer being set.

These methods have long been not actually needed, so their implementations are now removed and there is
no longer a need for an early return.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKWindowVisibilityObserver startObserving:]):
(-[WKWindowVisibilityObserver stopObserving:]):
(-[WKWindowVisibilityObserver _windowWillEnterFullScreen:]):
(-[WKWindowVisibilityObserver _windowWillExitFullScreen:]):
(WebKit::WebViewImpl::windowWillEnterFullScreen):
(WebKit::WebViewImpl::windowWillExitFullScreen):
When entering or exiting full screen, we now listen to their respective notifications.

Canonical link: https://commits.webkit.org/266273@main
